batch/batch/aioclient.py

```diff
@@ -66,26 +82,71 @@ def __init__(self, batch, job_id, attributes=None, parent_ids=None, _status=None
                 i = i + 1
 
     async def log(self):
-        return await self.batch._client._get(f'/batches/{self.batch_id}/jobs/{self.job_id}/log')
+        if self._batch.id is None:
+            raise ValueError("cannot get the log of a unsubmitted job")
+        return await self._batch._client._get(f'/batches/{self.batch_id}/jobs/{self.job_id}/log')
 
 
 class Batch:
     def __init__(self, client, id, attributes):
         self._client = client
         self.id = id
         self.attributes = attributes
+
+    async def cancel(self):
+        await self._client._patch(f'/batches/{self.id}/cancel')
+
+    async def status(self):
+        return await self._client._get(f'/batches/{self.id}')
+
+    async def wait(self):
+        i = 0
+        while True:
+            status = await self.status()
+            if status['complete']:
+                return status
+            j = random.randrange(math.floor(1.1 ** i))
+            time.sleep(0.100 * j)
+            # max 4.45s
+            if i < 64:
+                i = i + 1
+
+    async def delete(self):
+        await self._client._delete(f'/batches/{self.id}')
+
+
+class BatchBuilder:
+    def __init__(self, client, attributes, callback):
+        doc = {}
+        if attributes:
+            doc['attributes'] = attributes
+        if callback:
+            doc['callback'] = callback
+
+        self._client = client
+        self._batch = Batch(client, None, attributes)
```

Comment:
Slightly uncomfortable with this. I think you should instantiate Batch only when it has an `id`; otherwise it's kind of a useless container of data (none of the methods it defines will work, since they're all predicated on the id).

Reply:
Probably not blocking, and I assume this goes away in the single submission interface.